src/OrchardCore.Modules/OrchardCore.Autoroute/Indexes/AutoroutePartIndex.cs

```diff
@@ -97,19 +101,37 @@ public void Describe(DescribeContext<ContentItem> context)
                         return null;
                     }
 
+                    // Lazy initialization because of ISession cyclic dependency
+                    _contentDefinitionManager = _contentDefinitionManager ?? _serviceProvider.GetRequiredService<IContentDefinitionManager>();
+
+                    // Search for AutoroutePart
+                    var contentTypeDefinition = _contentDefinitionManager.GetTypeDefinition(contentItem.ContentType);
+
+                    // Validate that the content definition contains an AutoroutePart.
+                    // This prevents indexing parts that have been removed from the type definition, but are still present in the elements.
+                    var removedFromTypeDefinition = false;
+                    if (contentTypeDefinition == null || !contentTypeDefinition.Parts.Any(ctpd => ctpd.Name == nameof(AutoroutePart)))
+                    {
+                        // When the part has been removed enlist an update for after the session has been commited.
+                        removedFromTypeDefinition = true;
+                        contentItem.Remove<AutoroutePart>();
+                        var autorouteEntries = _serviceProvider.GetRequiredService<IAutorouteEntries>();
+                        await autorouteEntries.UpdateEntriesAsync();
+                    }
```

Comment:
@sebastienros I fixed the issue we had where an `AutoroutePart` continued life, and remained routable, even when it was removed from the type definition, and the content item was resaved, instead of using `When`.

Kept `When` for welded types, like `Layer` 

@jtkech here it will still remain in the index with a null path, until the next save. And enlists `UpdateEntriesAsync` so it will be removed from the route entries.
Haven't done the `LocalizedContentItemIndex` yet - different because it has no concept of a null path. So need to check with you on the best approach. 

But we can figure that out on another pr, if adding `Remove` is a good idea.

Reply:
@deanmarcussen 

When i worked on `AutoroutePartIndexProvider` i saw that an AutoroutePart remained routable even if removed from the definition, it was already the case before and as i remember it was okay after restarting the app, so too lazy to fix it ;)

So if i understand, in that case the autoroute part handler is not called, that's why we update the entries here. Updating entries is just to anticipate the refresh that would be done otherwise on the next request. So here maybe too much, hmm but why not, as it is done in the handler.

Hmm, because here the index provider is also a `ContentHandlerBase`, maybe good to do the job in `.UpdatedAsync()`, including refresh entries, and then enlist items in a `_partRemoved` hashset.

For the `LocalizedContentItemIndex`, yes we will see, maybe set the `LocalizationSet` and / or `Culture` to null.

Reply:
@jtkech done too quickly, just to try something.

Think it needs more thought, and design.

The `ContentHandlerBase` doesn't get called, because the part has been removed.

Perhaps if we want to clean these, a `ContentHandler` that checks against type definitions might be better.

Though I also lean towards a `CleanUp` style recipe, which does an export, with that check and remove, and allows you to recreate without.

There's something more explicit about it, and less dangerous, than just removing things because they don't match definitions, and also less headache to support in code.

I also think it's maybe more a development concern than a production concern. By the time you've got to production you've got templates that rely on these parts, and all sorts of other things to take into consideration when changing type definitions.

I'll keep this open for a bit, to keep the ideas going.

Reply:
@deanmarcussen 

> done too quickly, just to try something. Think it needs more thought, and design.

No, i think it's a good idea and seems to be almost ready

> The ContentHandlerBase doesn't get called, because the part has been removed.

I think you mean that the `IContentPartHandler` is not called, because `IContentHandler` is called

> Perhaps if we want to clean these, a ContentHandler that checks against type definitions might be better.

**I suggested to still do it in `AutoroutePartIndexProvider` as it is already a `ContentHandlerBase`**

About the other comments, hmm make sense for me to remove the part if it doesn't exist anymore in the definition, this when saving again the content item, or maybe being able to mark a part as soft deleted, allowing a future tool to retrieve it, but as said for now LGTM.

Reply:
> I think you mean that the IContentPartHandler is not called, because IContentHandler is called

You're totally right, done too quickly (and the idea of Removing was a later one), I'll move it there.

Reply:
> I suggested to still do it in AutoroutePartIndexProvider as it is already a ContentHandlerBase

I will tweak this when I have a moment, as I still agree with you @jtkech it is just merged before I got to it ;)

Reply:
@deanmarcussen Okay cool, no problem ;)